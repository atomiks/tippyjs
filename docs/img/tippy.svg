<?xml version="1.0" encoding="UTF-8"?>
<svg width="222px" height="184px" viewBox="0 0 222 184" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 41.2 (35397) - http://www.bohemiancoding.com/sketch -->
    <title>Artboard</title>
    <desc>Created with Sketch.</desc>
    <defs>
        <rect id="path-1" x="90.7780977" y="135.681765" width="40" height="40" rx="5"></rect>
        <mask id="mask-2" maskContentUnits="userSpaceOnUse" maskUnits="objectBoundingBox" x="0" y="0" width="40" height="40" fill="white">
            <use xlink:href="#path-1"></use>
        </mask>
        <linearGradient x1="6.21993923%" y1="1.53665412%" x2="96.5620339%" y2="30.4905087%" id="linearGradient-3">
            <stop stop-color="#A458FF" offset="0%"></stop>
            <stop stop-color="#FFA887" offset="18.4490593%"></stop>
            <stop stop-color="#FFDA90" offset="18.6084981%"></stop>
            <stop stop-color="#FFFFFF" offset="53.7527902%"></stop>
            <stop stop-color="#73FFD5" offset="53.7941586%"></stop>
            <stop stop-color="#78FFD6" offset="55.7374589%"></stop>
            <stop stop-color="#EDFFF6" offset="100%"></stop>
            <stop stop-color="#96FFEF" offset="100%"></stop>
        </linearGradient>
        <rect id="path-4" x="0.0116328814" y="0.147055547" width="221.53293" height="157.580832" rx="20"></rect>
        <mask id="mask-5" maskContentUnits="userSpaceOnUse" maskUnits="objectBoundingBox" x="0" y="0" width="221.53293" height="157.580832" fill="white">
            <use xlink:href="#path-4"></use>
        </mask>
    </defs>
    <g id="Page-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="Artboard">
            <use id="Rectangle-3" stroke="#253B59" mask="url(#mask-2)" stroke-width="20" fill="#A1FFEE" transform="translate(110.778098, 155.681765) rotate(-315.000000) translate(-110.778098, -155.681765) " xlink:href="#path-1"></use>
            <use id="Rectangle" stroke="#253B59" mask="url(#mask-5)" stroke-width="20" fill="url(#linearGradient-3)" xlink:href="#path-4"></use>
            <ellipse id="Oval" fill="#243A59" cx="76.3220539" cy="78.8703424" rx="9.00019972" ry="8.90879267"></ellipse>
            <ellipse id="Oval" fill="#243A59" cx="145.234142" cy="78.8703424" rx="9.00019972" ry="8.90879267"></ellipse>
            <path d="M9.30051464,21.4264399 C9.30051464,18.5526104 11.0565568,14.6911723 13.2181501,12.8169312 C13.2181501,12.8169312 13.4934787,12.5583429 13.8003903,12.325077 C14.1073018,12.0918111 15.565069,11.1156541 15.565069,11.1156541 C17.8501321,9.56995953 19.7025418,10.551332 19.7025418,13.3266137 L19.7025418,144.650037 C19.7025418,147.416811 17.7825275,148.508293 15.4115838,147.086439 L13.7571778,146.094292 C11.2958307,144.618223 9.30051464,141.091533 9.30051464,138.218104 L9.30051464,21.4264399 Z" id="Rectangle" fill="#FFFFFF"></path>
            <path d="M87.1387933,103.247251 C87.1387933,103.247251 107.917826,122.782037 134.417402,103.247251" id="Path-2" stroke="#253B59" stroke-width="10" stroke-linecap="round"></path>
        </g>
    </g>
</svg>