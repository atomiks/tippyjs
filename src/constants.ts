export const PASSIVE = {passive: true};

export const ROUND_ARROW =
  '<svg viewBox="0 0 18 7" xmlns="http://www.w3.org/2000/svg"><path d="M0 7s2.021-.015 5.253-4.218C6.584 1.051 7.797.007 9 0c1.203-.007 2.416 1.035 3.761 2.782C16.012 7.005 18 7 18 7H0z"/></svg>';

export const IOS_CLASS = `__NAMESPACE_PREFIX__-iOS`;
export const POPPER_CLASS = `__NAMESPACE_PREFIX__-popper`;
export const TOOLTIP_CLASS = `__NAMESPACE_PREFIX__-tooltip`;
export const CONTENT_CLASS = `__NAMESPACE_PREFIX__-content`;
export const BACKDROP_CLASS = `__NAMESPACE_PREFIX__-backdrop`;
export const ARROW_CLASS = `__NAMESPACE_PREFIX__-arrow`;
export const SVG_ARROW_CLASS = `__NAMESPACE_PREFIX__-svg-arrow`;

export const POPPER_SELECTOR = `.${POPPER_CLASS}`;
export const TOOLTIP_SELECTOR = `.${TOOLTIP_CLASS}`;
export const CONTENT_SELECTOR = `.${CONTENT_CLASS}`;
export const BACKDROP_SELECTOR = `.${BACKDROP_CLASS}`;
export const ARROW_SELECTOR = `.${ARROW_CLASS}`;
export const SVG_ARROW_SELECTOR = `.${SVG_ARROW_CLASS}`;
