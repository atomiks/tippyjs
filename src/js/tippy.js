import { version } from '../../package.json'

import { browser } from './core/globals'
import defaults from './core/defaults'

import isObjectLiteral from './utils/isObjectLiteral'
import getArrayOfElements from './utils/getArrayOfElements'
import polyfillVirtualReferenceProps from './utils/polyfillVirtualReferenceProps'

import createTooltips from './core/createTooltips'
import bindEventListeners from './core/bindEventListeners'

let eventListenersBound = false
let useCapture = false

/**
 * Exported module
 * @param {String|Element|Element[]|NodeList|Object} selector
 * @param {Object} options
 * @param {Boolean} one - create one tooltip
 * @return {Object}
 */
function tippy(selector, options, one) {
  if (browser.supported && !eventListenersBound) {
    bindEventListeners(useCapture)
    eventListenersBound = true
  }

  if (isObjectLiteral(selector)) {
    polyfillVirtualReferenceProps(selector)
  }

  options = { ...defaults, ...options }

  const references = getArrayOfElements(selector)
  const firstReference = references[0]

  return {
    selector,
    options,
    tooltips: browser.supported
      ? createTooltips(
          one && firstReference ? [firstReference] : references,
          options
        )
      : [],
    destroyAll() {
      this.tooltips.forEach(tooltip => tooltip.destroy())
      this.tooltips = []
    }
  }
}

tippy.version = version
tippy.browser = browser
tippy.defaults = defaults
tippy.one = (selector, options) => tippy(selector, options, true).tooltips[0]
tippy.disableAnimations = () => {
  defaults.updateDuration = defaults.duration = 0
  defaults.animateFill = false
}
tippy.useCapture = () => {
  useCapture = true
}

export default tippy
