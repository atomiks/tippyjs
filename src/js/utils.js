import { arrayFrom } from './ponyfills'

/**
 * Determines if a value is a plain object
 * @param {any} value
 * @return {Boolean}
 */
export function isPlainObject(value) {
  return {}.toString.call(value) === '[object Object]'
}

/**
 * Safe .hasOwnProperty check, for prototype-less objects
 * @param {Object} obj
 * @param {String} key
 * @return {Boolean}
 */
export function hasOwnProperty(obj, key) {
  return {}.hasOwnProperty.call(obj, key)
}

/**
 * Determines if a value is numeric
 * @param {any} value
 * @return {Boolean}
 */
export function isNumeric(value) {
  return !isNaN(value) && !isNaN(parseFloat(value))
}

/**
 * Returns an array of elements based on the value
 * @param {any} value
 * @return {Array}
 */
export function getArrayOfElements(value) {
  if (value instanceof Element || isPlainObject(value)) {
    return [value]
  }
  if (value instanceof NodeList) {
    return arrayFrom(value)
  }
  if (Array.isArray(value)) {
    return value
  }

  try {
    return arrayFrom(document.querySelectorAll(value))
  } catch (e) {
    return []
  }
}

/**
 * Returns a value at a given index depending on if it's an array or number
 * @param {any} value
 * @param {Number} index
 * @param {any} defaultValue
 */
export function getValue(value, index, defaultValue) {
  if (Array.isArray(value)) {
    const v = value[index]
    return v == null ? defaultValue : v
  }
  return value
}

/**
 * Focuses an element while preventing a scroll jump if it's not within the
 * viewport
 * @param {Element} el
 */
export function focus(el) {
  const x = window.scrollX || window.pageXOffset
  const y = window.scrollY || window.pageYOffset
  el.focus()
  scroll(x, y)
}

/**
 * Defers a function's execution until the call stack has cleared
 * @param {Function} fn
 */
export function defer(fn) {
  setTimeout(fn, 1)
}

/**
 * Debounce utility
 * @param {Function} fn
 * @param {Number} ms
 */
export function debounce(fn, ms) {
  let timeoutId
  return function() {
    clearTimeout(timeoutId)
    timeoutId = setTimeout(() => fn.apply(this, arguments), ms)
  }
}

/**
 * Prevents errors from being thrown while accessing nested modifier objects
 * in `popperOptions`
 * @param {Object} obj
 * @param {String} key
 * @return {Object}
 */
export function getModifier(obj, key) {
  return obj && obj.modifiers ? obj.modifiers[key] : {}
}
