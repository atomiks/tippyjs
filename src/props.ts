import {DefaultProps, Plugin, Props, ReferenceElement, Tippy} from './types';
import {
  hasOwnProperty,
  removeProperties,
  invokeWithArgsOrReturn,
} from './utils';
import {warnWhen} from './validation';

const pluginProps = {
  animateFill: false,
  followCursor: false,
  inlinePositioning: false,
  sticky: false,
};

const renderProps = {
  allowHTML: false,
  animation: 'fade',
  arrow: true,
  content: '',
  inertia: false,
  maxWidth: 350,
  role: 'tooltip',
  theme: '',
  zIndex: 9999,
};

export const defaultProps: DefaultProps = {
  appendTo: () => document.body,
  aria: {
    content: 'auto',
    expanded: 'auto',
  },
  delay: 0,
  duration: [300, 250],
  getReferenceClientRect: null,
  hideOnClick: true,
  ignoreAttributes: false,
  interactive: false,
  interactiveBorder: 2,
  interactiveDebounce: 0,
  moveTransition: '',
  offset: [0, 10],
  onAfterUpdate() {},
  onBeforeUpdate() {},
  onCreate() {},
  onDestroy() {},
  onHidden() {},
  onHide() {},
  onMount() {},
  onShow() {},
  onShown() {},
  onTrigger() {},
  onUntrigger() {},
  onClickOutside() {},
  placement: 'top',
  plugins: [],
  popperOptions: {},
  render: null,
  showOnCreate: false,
  touch: true,
  trigger: 'mouseenter focus',
  triggerTarget: null,
  ...pluginProps,
  ...renderProps,
};

const defaultKeys = Object.keys(defaultProps);

export const setDefaultProps: Tippy['setDefaultProps'] = (partialProps) => {
  /* istanbul ignore else */
  if (__DEV__) {
    validateProps(partialProps, []);
  }

  const keys = Object.keys(partialProps) as Array<keyof DefaultProps>;
  keys.forEach((key) => {
    (defaultProps as any)[key] = partialProps[key];
  });
};

export function getExtendedPassedProps(
  passedProps: Partial<Props> & Record<string, unknown>
): Partial<Props> {
  const plugins = passedProps.plugins || [];
  const pluginProps = plugins.reduce<Record<string, unknown>>((acc, plugin) => {
    const {name, defaultValue} = plugin;

    if (name) {
      acc[name] =
        passedProps[name] !== undefined ? passedProps[name] : defaultValue;
    }

    return acc;
  }, {});

  return {
    ...passedProps,
    ...pluginProps,
  };
}

export function getDataAttributeProps(
  reference: ReferenceElement,
  plugins: Plugin[]
): Record<string, unknown> {
  const propKeys = plugins
    ? Object.keys(getExtendedPassedProps({...defaultProps, plugins}))
    : defaultKeys;

  const props = propKeys.reduce(
    (acc: Partial<Props> & Record<string, unknown>, key) => {
      const valueAsString = (
        reference.getAttribute(`data-tippy-${key}`) || ''
      ).trim();

      if (!valueAsString) {
        return acc;
      }

      if (key === 'content') {
        acc[key] = valueAsString;
      } else {
        try {
          acc[key] = JSON.parse(valueAsString);
        } catch (e) {
          acc[key] = valueAsString;
        }
      }

      return acc;
    },
    {}
  );

  return props;
}

export function evaluateProps(
  reference: ReferenceElement,
  props: Props
): Props {
  const out = {
    ...props,
    content: invokeWithArgsOrReturn(props.content, [reference]),
    ...(props.ignoreAttributes
      ? {}
      : getDataAttributeProps(reference, props.plugins)),
  };

  out.aria = {
    ...defaultProps.aria,
    ...out.aria,
  };

  out.aria = {
    expanded:
      out.aria.expanded === 'auto' ? props.interactive : out.aria.expanded,
    content:
      out.aria.content === 'auto'
        ? props.interactive
          ? null
          : 'describedby'
        : out.aria.content,
  };

  return out;
}

export function validateProps(
  partialProps: Partial<Props> = {},
  plugins: Plugin[] = []
): void {
  const keys = Object.keys(partialProps) as Array<keyof Props>;
  keys.forEach((prop) => {
    const nonPluginProps = removeProperties(
      defaultProps,
      Object.keys(pluginProps)
    );

    let didPassUnknownProp = !hasOwnProperty(nonPluginProps, prop);

    // Check if the prop exists in `plugins`
    if (didPassUnknownProp) {
      didPassUnknownProp =
        plugins.filter((plugin) => plugin.name === prop).length === 0;
    }

    warnWhen(
      didPassUnknownProp,
      [
        `\`${prop}\``,
        "is not a valid prop. You may have spelled it incorrectly, or if it's",
        'a plugin, forgot to pass it in an array as props.plugins.',
        '\n\n',
        'All props: https://atomiks.github.io/tippyjs/v6/all-props/\n',
        'Plugins: https://atomiks.github.io/tippyjs/v6/plugins/',
      ].join(' ')
    );
  });
}
