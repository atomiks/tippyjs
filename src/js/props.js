import { getDataAttributeOptions } from './reference'
import { hasOwnProperty } from './utils'

/**
 * Evaluates options
 * @param {Element} reference
 * @param {Object} options
 * @return {Object}
 */
export function evaluateProps(reference, props) {
  const out = {
    ...props,
    ...(props.performance ? {} : getDataAttributeOptions(reference))
  }

  if (out.arrow) {
    out.animateFill = false
  }

  if (typeof out.appendTo === 'function') {
    out.appendTo = props.appendTo(reference)
  }

  if (typeof out.content === 'function') {
    out.content = props.content(reference)
  }

  return out
}

/**
 * Validates an object of options with the valid default props object
 * @param {Object} options
 * @param {Object} defaults
 */
export function validateOptions(options = {}, defaults) {
  Object.keys(options).forEach(option => {
    if (!hasOwnProperty(defaults, option)) {
      throw new Error(`[tippy]: \`${option}\` is not a valid option`)
    }
  })
}
