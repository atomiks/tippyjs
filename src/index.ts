import {version} from '../package.json';
import {defaultProps} from './props';
import createTippy from './createTippy';
import bindGlobalEventListeners, {
  currentInput,
} from './bindGlobalEventListeners';
import {
  arrayFrom,
  getArrayOfElements,
  isReferenceElement,
  isElement,
} from './utils';
import {warnWhen, validateTargets, validateProps} from './validation';
import {POPPER_SELECTOR} from './constants';
import {
  Props,
  Instance,
  Targets,
  PopperElement,
  HideAllOptions,
  Plugin,
  Tippy,
} from './types';

/**
 * Exported module
 */
function tippy(
  targets: Targets,
  optionalProps?: Partial<Props>,
  plugins: Plugin[] = [],
): Instance | Instance[] {
  if (__DEV__) {
    validateTargets(targets);
    validateProps(optionalProps, plugins);
  }

  bindGlobalEventListeners();

  const props: Props = {...defaultProps, ...optionalProps};

  const elements = getArrayOfElements(targets);

  if (__DEV__) {
    const isSingleContentElement = isElement(props.content);
    const isMoreThanOneReferenceElement = elements.length > 1;
    warnWhen(
      isSingleContentElement && isMoreThanOneReferenceElement,
      `tippy() was passed an Element as the \`content\` prop, but more than one
      tippy instance was created by this invocation. This means the content
      element will only be appended to the last tippy instance.
      
      Instead, pass the .innerHTML of the element, or use a function that
      returns a cloned version of the element instead.
      
      1) content: () => element.cloneNode(true)
      2) content: element.innerHTML`,
    );
  }

  const instances = elements.reduce<Instance[]>(
    (acc, reference): Instance[] => {
      const instance = reference && createTippy(reference, props, plugins);

      if (instance) {
        acc.push(instance);
      }

      return acc;
    },
    [],
  );

  return isElement(targets) ? instances[0] : instances;
}

tippy.version = version;
tippy.defaultProps = defaultProps;
tippy.setDefaultProps = setDefaultProps;
tippy.currentInput = currentInput;

/**
 * Mutates the defaultProps object by setting the props specified
 */
function setDefaultProps(partialProps: Partial<Props>): void {
  if (__DEV__) {
    validateProps(partialProps, []);
  }

  Object.keys(partialProps).forEach((key): void => {
    defaultProps[key] = partialProps[key];
  });
}

/**
 * Returns a proxy wrapper function that passes the plugins
 */
export function createTippyWithPlugins(outerPlugins: Plugin[]): Tippy {
  const fn = (
    targets: Targets,
    optionalProps?: Partial<Props>,
    innerPlugins: Plugin[] = [],
  ): Instance | Instance[] =>
    tippy(targets, optionalProps, [...outerPlugins, ...innerPlugins]);

  fn.version = version;
  fn.defaultProps = defaultProps;
  fn.setDefaultProps = setDefaultProps;
  fn.currentInput = currentInput;

  return fn;
}

/**
 * Hides all visible poppers on the document
 */
export function hideAll({
  exclude: excludedReferenceOrInstance,
  duration,
}: HideAllOptions = {}): void {
  arrayFrom(document.querySelectorAll(POPPER_SELECTOR)).forEach(
    (popper: PopperElement): void => {
      const instance = popper._tippy;

      if (instance) {
        let isExcluded = false;
        if (excludedReferenceOrInstance) {
          isExcluded = isReferenceElement(excludedReferenceOrInstance)
            ? instance.reference === excludedReferenceOrInstance
            : popper === excludedReferenceOrInstance.popper;
        }

        if (!isExcluded) {
          instance.hide(duration);
        }
      }
    },
  );
}

export default tippy;
