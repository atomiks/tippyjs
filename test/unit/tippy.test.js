import {h, cleanDocumentBody} from '../utils';

import {defaultProps, extraProps} from '../../src/props';
import {POPPER_SELECTOR} from '../../src/constants';
import tippy, {hideAll, createTippyWithPlugins} from '../../src';
import {getFormattedMessage} from '../../src/validation';

jest.useFakeTimers();

afterEach(cleanDocumentBody);

describe('tippy', () => {
  it('returns the expected object', () => {
    expect(typeof tippy(h())).toBe('object');
    expect(Array.isArray(tippy([h(), h()]))).toBe(true);
  });

  it('merges the default props with the supplied props', () => {
    expect(tippy(h(), {placement: 'bottom-end'}).props).toEqual({
      ...defaultProps,
      ...extraProps,
      placement: 'bottom-end',
    });
  });

  it('warns if invalid props(s) are supplied', () => {
    const spy = jest.spyOn(console, 'warn');

    tippy(h(), {placement: 'top', _someInvalidProp: true});

    expect(spy).toHaveBeenCalledTimes(1);

    spy.mockRestore();
  });

  it('handles falsy reference in an array', () => {
    tippy([null, false, 0, undefined]);
  });

  it('throws if passed falsy Target type', () => {
    expect(() => tippy(null)).toThrow();
    expect(() => tippy(undefined)).toThrow();
    expect(() => tippy(false)).toThrow();
    expect(() => tippy(0)).toThrow();
    expect(() => tippy(NaN)).toThrow();
    expect(() => tippy('')).toThrow();
  });

  it('warns if passed a single content element for many different references', () => {
    const spy = jest.spyOn(console, 'warn');
    const targets = [h(), h()];

    tippy(targets, {content: document.createElement('div')});

    expect(spy).toHaveBeenCalledWith(
      ...getFormattedMessage(
        `tippy() was passed an Element as the \`content\` prop, but more than one
      tippy instance was created by this invocation. This means the content
      element will only be appended to the last tippy instance.
      
      Instead, pass the .innerHTML of the element, or use a function that
      returns a cloned version of the element instead.
      
      1) content: () => element.cloneNode(true)
      2) content: element.innerHTML`,
      ),
    );

    spy.mockRestore();
  });
});

describe('tippy.setDefaultProps()', () => {
  it('changes the default props applied to instances', () => {
    const newPlacement = 'bottom-end';

    tippy.setDefaultProps({placement: newPlacement});

    expect(defaultProps.placement).toBe(newPlacement);
  });

  it('is validated', () => {
    const spy = jest.spyOn(console, 'warn');

    tippy.setDefaultProps({theme: 'google'});

    expect(spy).toHaveBeenCalledWith(
      ...getFormattedMessage(
        `The included theme "google" was renamed to "material" in v5.`,
      ),
    );

    spy.mockRestore();
  });
});

describe('hideAll()', () => {
  it('hides all tippys on the document, ignoring `hideOnClick`', () => {
    const props = {showOnCreate: true, hideOnClick: false};
    const instances = [...Array(3)].map(() => tippy(h(), props));

    jest.runAllTimers();
    hideAll();

    instances.forEach(instance => {
      expect(instance.state.isVisible).toBe(false);
    });
  });

  it('respects `duration` option', () => {
    const props = {showOnCreate: true, duration: 100};
    const instances = [...Array(3)].map(() => tippy(h(), props));

    jest.runAllTimers();
    hideAll({duration: 0});

    instances.forEach(instance => {
      expect(instance.state.isMounted).toBe(false);
    });
  });

  it('respects `exclude` option', () => {
    const props = {showOnCreate: true};
    const instances = [...Array(3)].map(() => tippy(h(), props));

    jest.runAllTimers();
    hideAll({exclude: instances[0]});

    instances.forEach(instance => {
      expect(instance.state.isVisible).toBe(
        instance === instances[0] ? true : false,
      );
    });
  });

  it('respects `exclude` option as type ReferenceElement for multiple tippys', () => {
    const props = {showOnCreate: true, multiple: true};
    const ref = h();

    tippy(ref, props);
    tippy(ref, props);
    hideAll({exclude: ref});

    const instances = [...document.querySelectorAll(POPPER_SELECTOR)].map(
      popper => popper._tippy,
    );

    instances.forEach(instance => {
      expect(instance.state.isVisible).toBe(true);
    });
  });
});

describe('createTippyWithPlugins', () => {
  it('returns a higher order tippy function that passes plugins', () => {
    const plugin = {fn: () => ({})};
    const tippy = createTippyWithPlugins([plugin]);
    const instance = tippy(document.createElement('div'));

    expect(instance.plugins).toEqual([plugin]);
  });

  it('preserves statics', () => {
    const plugin = {fn: () => ({})};
    const tippy = createTippyWithPlugins([plugin]);

    expect(tippy.version).toBeDefined();
    expect(tippy.defaultProps).toBeDefined();
    expect(tippy.setDefaultProps).toBeDefined();
    expect(tippy.currentInput).toBeDefined();
  });

  it('warns', () => {
    const spy = jest.spyOn(console, 'warn');

    createTippyWithPlugins([]);

    expect(spy).toHaveBeenCalledWith(
      ...getFormattedMessage(
        `createTippyWithPlugins([...]) has been deprecated.

      Use tippy.setDefaultProps({plugins: [...]}) instead.`,
      ),
    );

    spy.mockRestore();
  });
});
