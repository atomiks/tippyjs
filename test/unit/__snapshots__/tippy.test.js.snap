// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tippy merges the default props with the supplied props 1`] = `
Object {
  "allowHTML": false,
  "animateFill": false,
  "animation": "fade",
  "appendTo": [Function],
  "aria": Object {
    "content": "describedby",
    "expanded": false,
  },
  "arrow": true,
  "content": "__DEFAULT_TEST_CONTENT__",
  "delay": 0,
  "duration": 0,
  "followCursor": false,
  "getReferenceClientRect": null,
  "hideOnClick": true,
  "ignoreAttributes": false,
  "inertia": false,
  "inlinePositioning": false,
  "interactive": false,
  "interactiveBorder": 2,
  "interactiveDebounce": 0,
  "maxWidth": 350,
  "moveTransition": "",
  "offset": Array [
    0,
    10,
  ],
  "onAfterUpdate": [Function],
  "onBeforeUpdate": [Function],
  "onClickOutside": [Function],
  "onCreate": [Function],
  "onDestroy": [Function],
  "onHidden": [Function],
  "onHide": [Function],
  "onMount": [Function],
  "onShow": [Function],
  "onShown": [Function],
  "onTrigger": [Function],
  "onUntrigger": [Function],
  "placement": "bottom-end",
  "plugins": Array [],
  "popperOptions": Object {},
  "render": [Function],
  "role": "tooltip",
  "showOnCreate": false,
  "sticky": false,
  "theme": "",
  "touch": true,
  "trigger": "mouseenter focus",
  "triggerTarget": null,
  "zIndex": 9999,
}
`;
